packages/synapse-react-client/src/components/ChangePassword/useChangePasswordFormState.tsx

```diff
@@ -0,0 +1,146 @@
+import {
+  ChangePasswordWithCurrentPassword,
+  ChangePasswordWithToken as ChangePasswordWithTokenObject,
+  ChangePasswordWithTwoFactorAuthToken,
+  PasswordResetSignedToken,
+  TwoFactorAuthErrorResponse,
+  TwoFactorAuthOtpType,
+} from '@sage-bionetworks/synapse-types'
+import { useChangePassword } from '../../synapse-queries'
+import { Alert, Typography } from '@mui/material'
+import {
+  RECOVERY_CODE_GUIDANCE_TEXT_SHORT,
+  TOTP_GUIDANCE_TEXT,
+} from '../Authentication/Constants'
+import { ONE_TIME_PASSWORD_STEP, OneTimePasswordForm } from '../Authentication'
+
+export type UseChangePasswordFormStateOptions = {
+  onChangePasswordSuccess?: () => void
+}
+
+/**
+ * Hook that handles submitting a change password request and prompting the user for 2FA if necessary.
+ * @param options
+ */
+export default function useChangePasswordFormState(
+  options: UseChangePasswordFormStateOptions,
+) {
+  // Store new password in state so that we can re-use it if 2FA is required
+  const [newPassword, setNewPassword] = useState<string>('')
+  const [twoFactorAuthErrorResponse, setTwoFactorAuthErrorResponse] = useState<
+    TwoFactorAuthErrorResponse | undefined
+  >()
+  const [otpStep, setOtpStep] =
+    useState<ONE_TIME_PASSWORD_STEP>('VERIFICATION_CODE')
+  const {
+    mutate: changePassword,
+    isPending,
+    error,
+  } = useChangePassword({
+    onSuccess: maybeTwoFactorResponse => {
+      if (maybeTwoFactorResponse) {
+        // We store the 2FA Error Response in state because the `changePasswordResult` may change before we are done with the 2FA code,
+        // e.g. if the user enters the wrong 2FA code
+        setTwoFactorAuthErrorResponse(maybeTwoFactorResponse)
+      } else {
+        setTwoFactorAuthErrorResponse(undefined)
+        if (options.onChangePasswordSuccess) {
+          options.onChangePasswordSuccess()
+        }
+      }
+    },
+  })
+
+  function handleChangePasswordWithCurrentPassword(
+    username: string,
+    currentPassword: string,
+    newPassword: string,
+  ) {
+    setNewPassword(newPassword)
+    const changeRequest: ChangePasswordWithCurrentPassword = {
+      username,
+      currentPassword,
+      newPassword,
+      concreteType:
+        'org.sagebionetworks.repo.model.auth.ChangePasswordWithCurrentPassword',
+    }
+    changePassword(changeRequest)
+  }
+
+  function handleChangePasswordWithResetToken(
+    newPassword: string,
+    passwordChangeToken: PasswordResetSignedToken,
+  ) {
+    setNewPassword(newPassword)
+    const changeRequest: ChangePasswordWithTokenObject = {
+      newPassword,
+      concreteType:
+        'org.sagebionetworks.repo.model.auth.ChangePasswordWithToken',
+      passwordChangeToken: passwordChangeToken,
+    }
+    changePassword(changeRequest)
+  }
+
+  function handleChangePasswordWithOtp(
+    newPassword: string,
+    code: string,
+    otpType: TwoFactorAuthOtpType,
+  ) {
+    if (twoFactorAuthErrorResponse) {
```

Comment:
`twoFactorAuthPrompt` could be included in a component without checking whether `promptForTwoFactorAuth` was true. If `twoFactorAuthResponse` was undefined, then submitting `OneTimePasswordForm` would result in `handleChangePasswordWithOtp` being a noop. Sems like an edge case (likely related to a mistake configuring a component to use this custom hook), but I'm wondering whether it would be useful to log a message to help the developer troubleshoot that case? Up to you though!

Reply:
Updated `twoFactorAuthPrompt`  to be a function component where a fragment is rendered if `promptForTwoFactorAuth` is used. `OneTimePasswordForm` will now never be shown if `promptForTwoFactorAuth` is false